File: backend/src/model/issue.go
package model

import (
	"time"
	"webservices/src/types"
)

type Issue struct {
	ID          string              `gorm:"primaryKey;type:uuid;default:uuid_generate_v4()" json:"id"`
	ProjectID   string              `gorm:"type:uuid;column:project_id" json:"projectId"`
	Title       string              `gorm:"not null" json:"title"`
	Type        types.IssueType     `gorm:"type:issue_type;default:'task'" json:"type"`
	Priority    types.IssuePriority `gorm:"type:issue_priority;default:'medium'" json:"priority"`
	Status      types.IssueStatus   `gorm:"type:issue_status;default:'todo'" json:"status"`
	AssigneeID  *string             `gorm:"type:uuid;column:assignee_id" json:"assigneeId,omitempty"`
	ReporterID  *string             `gorm:"type:uuid;column:reporter_id" json:"reporterId,omitempty"`
	CreatorID   *string             `gorm:"type:uuid;column:creator_id" json:"creatorId,omitempty"`
	StartDate   *time.Time          `gorm:"column:start_date" json:"startDate,omitempty"`
	DueDate     *time.Time          `gorm:"column:due_date" json:"dueDate,omitempty"`
	DoneDate    *time.Time          `gorm:"column:done_date" json:"doneDate,omitempty"`
	Label       *string             `json:"label,omitempty"`
	Description *string             `json:"description,omitempty"`
	Goal        *string             `json:"goal,omitempty"`
	Parents     *string             `json:"parents,omitempty"`
	Order       int                 `gorm:"column:order_index;default:0" json:"order"`
	CreatedAt   time.Time           `gorm:"column:created_at;default:now();<-:create" json:"createdAt"`
	UpdatedAt   time.Time           `gorm:"column:updated_at;autoUpdateTime" json:"updatedAt"`

	Project  Project `gorm:"foreignKey:ProjectID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"project,omitzero"`
	Assignee *User   `gorm:"foreignKey:AssigneeID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"assignee,omitempty"`
	Reporter *User   `gorm:"foreignKey:ReporterID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"reporter,omitempty"`
	Creator  *User   `gorm:"foreignKey:CreatorID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"creator,omitempty"`

	Comments   []Comment        `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;" json:"comments,omitempty"`
	Items      []IssueItem      `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;" json:"items,omitempty"`
	Activities []RecentActivity `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;" json:"activities,omitempty"`
}

func (Issue) TableName() string {
	return "issues"
}

File: frontend/src/types/schemas/issue.ts
import { z } from 'zod';
import { User } from '@/types/schemas/user';
import { Project } from '@/types/schemas/project';
import { Comment } from '@/types/schemas/comment';
import { Activity } from '@/types/schemas/activity';
import { IssueItem } from '@/types/schemas/issue-item';

import {
    IssueStatus,
    STATUS_MAP,
    IssuePriority,
    PRIORITY_MAP,
    IssueType,
    ISSUE_TYPE_MAP
} from '@/types/misc';

export const issueSchema = z.object({
    title: z.string().min(1, 'Title is required').max(60, 'Maximum 60 characters'),
    type: z.enum(
        Object.keys(ISSUE_TYPE_MAP) as [IssueType, ...IssueType[]],
        { message: 'Type value is not valid' }
    ),
    priority: z.enum(
        Object.keys(PRIORITY_MAP) as [IssuePriority, ...IssuePriority[]],
        { message: 'Priority value is not valid' }
    ),
    status: z.enum(
        Object.keys(STATUS_MAP) as [IssueStatus, ...IssueStatus[]],
        { message: 'Status value is not valid' }
    ),
    assigneeId: z.string().optional(),
    startDate: z.date().optional(),
    dueDate: z.date().optional(),
    label: z.string().optional(),
    reporterId: z.string().optional(),
    description: z.string().optional(),
    goal: z.string().optional(),
    parents: z.string().optional(),
});

export type MoveDirection = 'top' | 'up' | 'down' | 'bottom'
export type IssueSchema = z.infer<typeof issueSchema>

export interface Issue extends IssueSchema {
    id: string
    projectId: string
    parents?: string
    order: number
    creatorId?: string
    doneDate?: Date
    project?: Project
    assignee?: User
    reporter?: User
    creator?: User
    comments?: Comment[]
    items?: IssueItem[]
    createdAt: string
    updatedAt: string
    activities?: Activity[]
}